src/string_map.c

```diff
@@ -147,46 +172,34 @@ rcutils_string_map_reserve(rcutils_string_map_t * string_map, size_t capacity)
     return RCUTILS_RET_OK;
   } else if (capacity == 0) {
     // if the requested capacity is zero, then make sure the existing keys and values are free'd
-    allocator.deallocate(string_map->impl->keys, allocator.state);
-    string_map->impl->keys = NULL;
-    allocator.deallocate(string_map->impl->values, allocator.state);
-    string_map->impl->values = NULL;
+    allocator.deallocate(string_map->impl->key_value_pairs, allocator.state);
+    string_map->impl->key_value_pairs = NULL;
     // falls through to normal function end
   } else {
     // if the capacity non-zero and different, use realloc to increase/shrink the size
     // note that realloc when the pointer is NULL is the same as malloc
     // note also that realloc will shrink the space if needed
 
-    // ensure that reallocate won't overflow capacity
-    if (capacity > (SIZE_MAX / sizeof(char *))) {
+    // ensure that reallocate won't overflow SIZE_MAX
+    if (capacity > (SIZE_MAX / sizeof(key_value_pair_t))) {
```

Comment:
Why capacity is down into half? originally capacity is `(SIZE_MAX / sizeof(char *))` for both entries, but current code shrinks it down to half? i might be mistaken...

Reply:
You are correct; on a 64-bit platform, the old check was:
```
capacity > (SIZE_MAX / sizeof(char *)) === capacity > (18446744073709551615UL / 8) === capacity > 2305843009213693951
```

The new check would be:
```
capacity > (SIZE_MAX / sizeof(key_value_pair_t)) === capacity > (18446744073709551615UL / 16) === capacity > 1152921504606846975
```

On a 32-bit platform (which we don't currently support), the old would be:
```
capacity > (SIZE_MAX / sizeof(char *)) === capacity > (4294967295UL / 8) === capacity > 536870911
```

The new would be:
```
capacity > (SIZE_MAX / sizeof(char *)) === capacity > (4294967295UL / 16) === capacity > 268435455
```

In all cases, this is more-or-less just a check to see if someone allocated something crazy; we don't really expect maps with that many entries.  So I don't see a huge deal here, but if you have other thoughts on what we might change this to I'm happy to hear it.

Reply:
yeah, allocating this much is crazy enough, current code should do good for checking 👍 (this was more like question.)